src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -1149,18 +1150,34 @@ public static IEnumerable<Diagnostic> GetEffectiveDiagnostics(ImmutableArray<Dia
 
         private static IEnumerable<Diagnostic> GetEffectiveDiagnosticsImpl(ImmutableArray<Diagnostic> diagnostics, Compilation compilation)
         {
+            if (diagnostics.IsEmpty)
+            {
+                yield break;
+            }
+
             var suppressMessageState = new SuppressMessageAttributeState(compilation);
+            var semanticModelsByTree = new ConcurrentDictionary<SyntaxTree, SemanticModel>();
```

Comment:
>ConcurrentDictionary [](start = 43, length = 20)

Do we need a `ConcurrentDictionary<,>` rather than `Dictionary<,>`? (Is the resulting `IEnumerator<Diagnostic>` accessed on multiple threads? Is the enumerator implementation thread-safe?)

Reply:
Yeah, a dictionary would likely suffice, I was just being more defensive.